### PR TITLE
Add Test chaining support

### PR DESCRIPTION
**Pull request by mukut1994**

### CODE REVIEW

File: MK94.Assert.NUnit.Test/ChainingTests.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        [Test]
+        public void PreventChainingMethodNonNUnit()
+        {
+            // TODO add non-task overload to MatchesException<T>
+            DiskAssert.MatchesException<InvalidOperationException>("Exception non test method",
+                Task.Run(() => DiskAssert.WithInputs().From(() => { }))); 
+        }
+
+        [Test]
+        public void Step1()
```
Comment (mithileshz) [inline MK94.Assert.NUnit.Test/ChainingTests.cs]: TestCase1

File: MK94.Assert.NUnit.Test/ChainingTests.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            DiskAssert.Matches("Step 1", new TestObject { A = 1, B = 2, C = 3 });
+        }
+
+        [Test]
+        public void Step2()
```
Comment (mithileshz) [inline MK94.Assert.NUnit.Test/ChainingTests.cs]: TestCase2

File: MK94.Assert.NUnit.Test/ChainingTests.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+            context.A += 10;
+            context.B += 10;
+            context.C += 10;
+
+            DiskAssert.Matches("Step 1", context);
+        }
+
+        [Test]
+        public void Step3()
```
Comment (mithileshz) [inline MK94.Assert.NUnit.Test/ChainingTests.cs]: TestCase3

File: MK94.Assert.NUnit/TestChainExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace MK94.Assert.NUnit
+{
+    public static class TestChainExtensions
+    {
+        public static TestChainer From(this TestChainer chainer, Action step)
+        {
+            if (!step.Method.GetCustomAttributes(typeof(TestAttribute), false).Any())
+                throw new InvalidOperationException($@"This does not seem to be an NUnit test method.
+Call {nameof(TestChainer.From)}(string, string) instead to set it manually.");
```
Comment (mithileshz) [inline MK94.Assert.NUnit/TestChainExtensions.cs]: Could the (string, string) not be automated too? So we don't have to worry about updating this if the method signature changes?

File: MK94.Assert/DiskAsserter.cs [hunk 5]
```diff
@@ -38,31 +37,32 @@ public class DiskAsserter
         /// </summary>
         /// <param name="step">A descriptive name of the step that generated the data</param>
         /// <param name="rawData">The raw data to be compared</param>
+        /// <param name="formatter">The error message formatter to use when there are differences</param>
         /// <returns>The unmodified <paramref name="rawData"/></returns>
         /// <exception cref="Exception">Thrown when some differences have been detected</exception>
         public string MatchesRaw(string step, string rawData, string fileType = null, IDifferenceFormatter<string> formatter = null)
```
Comment (mithileshz) [inline MK94.Assert/DiskAsserter.cs:44]: fileType parameter is still missing from the comments descriving the method above.

File: MK94.Assert/Output/IFileOutput.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿using System.IO;
+
+namespace MK94.Assert.Output
+{
+	/// <summary>
+	/// Used for writing/reading files. <br />
+	/// E.g. write to disk/memory/network etc
+	/// </summary>
+    	public interface IFileOutput
```
Comment (mithileshz) [inline MK94.Assert/Output/IFileOutput.cs]: Indentation

File: MK94.Assert/Output/ITestOutput.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using System.IO;
+
+namespace MK94.Assert.Output
+{
+	/// <summary>
+	/// Used for writing and reading Test output data. <br />
+	/// Responsible for caching reads, physical file location and hashing.
+	/// </summary>
+    	public interface ITestOutput
```
Comment (mithileshz) [inline MK94.Assert/Output/ITestOutput.cs]: indentation

File: TestData/MK94.Assert.NUnit.Test.ChainingTests/PreventChainingMethodNonNUnit/Exception non test method [hunk 8]
```diff
@@ -0,0 +1,12 @@
+This does not seem to be an NUnit test method.
```
Comment (mithileshz) [inline TestData/MK94.Assert.NUnit.Test.ChainingTests/PreventChainingMethodNonNUnit/Exception non test method]: it not an `an` NUnit Test Method but a `a`

File: TestData/MK94.Assert.NUnit.Test.ChainingTests/PreventChainingMethodNonNUnit/Exception non test method [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+Call From(string, string) instead to set it manually.
```
Comment (mithileshz) [inline TestData/MK94.Assert.NUnit.Test.ChainingTests/PreventChainingMethodNonNUnit/Exception non test method:2]: In the future might want to direct to documentation instead?

**Review by mithileshz — CHANGES_REQUESTED** [review]

File: MK94.Assert.NUnit/TestChainExtensions.cs (hunk 4 above)
Comment (mukut1994) [reply MK94.Assert.NUnit/TestChainExtensions.cs]: Not that I know of

File: MK94.Assert/DiskAsserter.cs (hunk 5 above)
Comment (mukut1994) [reply MK94.Assert/DiskAsserter.cs:44]: Fixed

File: TestData/MK94.Assert.NUnit.Test.ChainingTests/PreventChainingMethodNonNUnit/Exception non test method (hunk 9 above)
Comment (mukut1994) [reply TestData/MK94.Assert.NUnit.Test.ChainingTests/PreventChainingMethodNonNUnit/Exception non test method:2]: Yea that's a good idea actually, I guess that's something to put in the backlog
Need to have the docs first after all

**Review by mithileshz — APPROVED** [review]

**Pull request merged**